core/drivers/clk/sam/sama7g5_clk.c

```diff
@@ -0,0 +1,1620 @@
+			"ethpll_divpmcck", },
+		.pp_mux_table = { 5, 7, 8, 9, 10, },
+		.pp_count = 5,
+		.pp_chg_id = INT_MIN,
+	},
+	{
+		.n  = "tcb1_ch0_gclk",
+		.id = 91,
+		.r = { .max = 200000000 },
+		.pp = { "syspll_divpmcck", "imgpll_divpmcck",
+			"baudpll_divpmcck", "audiopll_divck",
+			"ethpll_divpmcck", },
+		.pp_mux_table = { 5, 7, 8, 9, 10, },
+		.pp_count = 5,
+		.pp_chg_id = INT_MIN,
+	},
+	{
+		.n  = "tcpca_gclk",
+		.id = 94,
+		.r = { .max = 32768, },
+		.pp_chg_id = INT_MIN,
+	},
+	{
+		.n  = "tcpcb_gclk",
+		.id = 95,
+		.r = { .max = 32768, },
+		.pp_chg_id = INT_MIN,
+	},
+};
+
+/* MCK0 characteristics. */
+static const struct clk_master_charac mck0_characteristics = {
+	.output = { .min = 32768, .max = 200000000 },
+	.divisors = { 1, 2, 4, 3, 5 },
+	.have_div3_pres = 1,
+};
+
+/* MCK0 layout. */
+static const struct clk_master_layout mck0_layout = {
+	.mask = 0x773,
+	.pres_shift = 4,
+	.offset = 0x28,
+};
+
+/* Peripheral clock layout. */
+static const struct clk_pcr_layout sama7g5_pcr_layout = {
+	.offset = 0x88,
+	.cmd = BIT(31),
+	.div_mask = GENMASK_32(27, 20),
+	.gckcss_mask = GENMASK_32(12, 8),
+	.pid_mask = GENMASK_32(6, 0),
+};
+
+static const struct clk_programmable_layout sama7g5_prog_layout = {
+	.pres_mask = 0xff,
+	.pres_shift = 8,
+	.css_mask = GENMASK_32(4, 0),
+	.have_slck_mck = 0,
+	.is_pres_direct = 1,
+};
+
+static const struct {
+	const char *n;
+	uint8_t id;
+} sama7g5_progck[] = {
+	{ .n = "prog0", .id = 0 },
+	{ .n = "prog1", .id = 1 },
+	{ .n = "prog2", .id = 2 },
+	{ .n = "prog3", .id = 3 },
+	{ .n = "prog4", .id = 4 },
+	{ .n = "prog5", .id = 5 },
+	{ .n = "prog6", .id = 6 },
+	{ .n = "prog7", .id = 7 },
+};
+
+static struct pmc_data *sama7g5_pmc;
+
+vaddr_t at91_pmc_get_base(void)
+{
+	assert(sama7g5_pmc);
+
+	return sama7g5_pmc->base;
+}
+
+TEE_Result at91_pmc_clk_get(unsigned int type, unsigned int idx,
+			    struct clk **clk)
+{
+	return pmc_clk_get(sama7g5_pmc, type, idx, clk);
+}
+
+static TEE_Result pmc_setup_sama7g5(const void *fdt, int nodeoffset,
+				    const void *data __unused)
+{
+	struct clk *pll_frac_clk[PLL_ID_MAX] = {NULL};
+	struct clk *pll_div_clk[PLL_ID_MAX] = {NULL};
+	TEE_Result res = TEE_ERROR_GENERIC;
+	const uint32_t *fdt_prop = NULL;
+	struct pmc_clk *pmc_clk = NULL;
+	struct clk *parents[11] = {NULL};
+	struct clk *main_xtal_clk = NULL;
+	struct clk *main_rc_osc = NULL;
+	struct clk *main_osc = NULL;
+	struct clk *mck0_clk = NULL;
+	struct clk *main_clk = NULL;
+	struct clk *md_slck = NULL;
+	struct clk *td_slck = NULL;
+	struct clk *parent = NULL;
+	struct clk *clk = NULL;
+	unsigned int i = 0;
+	unsigned int j = 0;
+	vaddr_t base = 0;
+	size_t size = 0;
+	int bypass = 0;
+
+	if (dt_map_dev(fdt, nodeoffset, &base, &size, DT_MAP_AUTO) < 0)
+		panic();
+
+	if (fdt_get_status(fdt, nodeoffset) == DT_STATUS_OK_SEC)
+		matrix_configure_periph_secure(ID_PMC);
+
+	res = clk_dt_get_by_name(fdt, nodeoffset, "md_slck", &md_slck);
+	if (res)
+		panic();
```

Comment:
why not `retrun res` in case probe function is deferred?
Ditto for the 2 other calls right below.